zebra/zebra_nhg.c

```diff
@@ -1675,10 +1675,14 @@ int nexthop_active_update(struct route_node *rn, struct route_entry *re)
 		new_active =
 			nexthop_active_check(rn, re, nexthop);
 
-		if (new_active
-		    && nexthop_group_active_nexthop_num(&new_grp)
-			       >= zrouter.multipath_num) {
+		if (new_active && curr_active >= zrouter.multipath_num) {
 			UNSET_FLAG(nexthop->flags, NEXTHOP_FLAG_ACTIVE);
+
+			/* Set its resolved nexthop as inactive as well */
+			if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
+				UNSET_FLAG(nexthop->resolved->flags,
+					   NEXTHOP_FLAG_ACTIVE);
```

Comment:
do you need to travel all the way down the resolution list - is just the first level ok?

Reply:
We only allow one level of resolution at this time, but you're right probably worth adding that in case it changes.


```
                } else if (CHECK_FLAG(re->flags, ZEBRA_FLAG_ALLOW_RECURSION)) {  
                        resolved = 0;                                            
                        for (ALL_NEXTHOPS_PTR(match->ng, newhop)) {              
                                if (!CHECK_FLAG(match->status,                   
                                                ROUTE_ENTRY_INSTALLED))          
                                        continue;                                
                                if (!nexthop_valid_resolve(nexthop, newhop))     
                                        continue;                                
                                                                                 
                                SET_FLAG(nexthop->flags,                         
                                         NEXTHOP_FLAG_RECURSIVE);                
                                nexthop_set_resolved(afi, newhop, nexthop);      
                                resolved = 1;                                    
                        }  
```

```
/* Checks if nexthop we are trying to resolve to is valid */                     
static bool nexthop_valid_resolve(const struct nexthop *nexthop,                 
                                  const struct nexthop *resolved)                
{                                                                                
        /* Can't resolve to a recursive nexthop */                               
        if (CHECK_FLAG(resolved->flags, NEXTHOP_FLAG_RECURSIVE))                 
                return false;                                                    
                                                                                 
        switch (nexthop->type) {                                                 
        case NEXTHOP_TYPE_IPV4_IFINDEX:                                          
        case NEXTHOP_TYPE_IPV6_IFINDEX:                                          
                /* If the nexthop we are resolving to does not match the         
                 * ifindex for the nexthop the route wanted, its not valid.      
                 */                                                              
                if (nexthop->ifindex != resolved->ifindex)                       
                        return false;                                            
                break;                                                           
        case NEXTHOP_TYPE_IPV4:                                                  
        case NEXTHOP_TYPE_IPV6:                                                  
        case NEXTHOP_TYPE_IFINDEX:                                               
        case NEXTHOP_TYPE_BLACKHOLE:                                             
                break;                                                           
        }                                                                        
                                                                                 
        return true;                                                             
}  
```

We dont allow you to resolve to a resolved ^^^

Reply:
ah, yes, sorry - was misremembering what happens when we have bgp->igp->CONNECTED (for example)

Reply:
Changed it to loop until nexthop->resolved is NULL